### ISSUE TITLE
pytorch backend lstm very +10x slow (maybe batch size with pytorch backend has different semantics than the traditional Keras semantics?)

### ISSUE THREAD

**Issue by mw66** [keras-team/keras#20902]
https://stackoverflow.com/questions/78717341/keras-training-speed-with-pytorch-backend-is-a-lot-slower-than-with-tensorflow

"""
I am on native Windows and I used old Keras with TensorFlow 2.10 (GPU accelerated) before. I wanted to try Keras 3 with PyTorch backend. Can someone please help me why this model trains 10x slower with Keras 3.4.1 and PyTorch 2.3.1 backend? With my GPU a single epoch takes a little more than 2 minutes with TF, and over 20 minutes with PyTorch.

import os
os.environ["KERAS_BACKEND"] = "torch"
import torch
torch.cuda.is_available() # <-- returns True

import keras
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import Dropout
from keras.layers import LSTM
from keras import optimizers
from keras.regularizers import l2

x_train, y_train = np.float32(x_train), np.float32(y_train)
x_val, y_val = np.float32(x_val), np.float32(y_val)

model=Sequential()
reg=0.00001
model.add(LSTM( 80, return_sequences=True , dropout=0.0, kernel_regularizer=l2(reg), recurrent_regularizer=l2(reg), input_shape=(x_train.shape[1], x_train.shape[2]) ))
model.add(LSTM( 80, return_sequences=False, dropout=0.0, kernel_regularizer=l2(reg), recurrent_regularizer=l2(reg) ))
model.add(Dense(40))
model.add(Dense(40))
model.add(Dense(1))
opt = optimizers.Adam(learning_rate=lrate)
model.compile(optimizer=opt, loss='mean_squared_error')

from keras.callbacks import ModelCheckpoint
from keras.callbacks import BackupAndRestore
savecallback = ModelCheckpoint(basefolder+"/"+modelfile, save_best_only=False, monitor='val_loss', mode='min', verbose=1)
backupcallback = BackupAndRestore(basefolder+"/tmp/backup_"+modelfile)

hist=model.fit(x_train, y_train, validation_data=(x_val, y_val), batch_size=batchsize, epochs=20, callbacks=[savecallback, backupcallback])

I verified GPU acceleration with both backends.
"""

**Comment by mw66** [comment]
https://stackoverflow.com/a/79438138/873275

"""
Experiencing the same problem.

I noticed that with pytorch backend the GPU memory is ~10x smaller, so I increased the batch size to be 16x, so the training speed is 16x faster. Now comparable to the TensorFlow backend (however, the GPU utilization is still low, ~3% vs ~30% with TF).

NOTE: increasing the batch size may affect training quality, which is yet to be compared.

I suspect batch size with pytorch backend has different semantics than the traditional Keras semantics. See here: https://discuss.pytorch.org/t/solved-pytorch-lstm-50x-slower-than-keras-tf-cudnnlstm/10043/8
"""

**Comment by mw66** [comment]
Related issues: "the LSTM and GRU are considerably (several times) slower"

https://github.com/keras-team/keras/issues/19177#issue-2134537516

**Comment by sonali-kumari1** [comment]
Hi @mw66 -

Sorry for the delayed response. I have tested your code with latest version of keras(3.9.2) in this [gist](https://colab.sandbox.google.com/gist/sonali-kumari1/559ceb0b4a645572dd69790949fe4b54/20931.ipynb). On a GPU setup, it takes approximately 1 minute per epoch, resulting in a total training time of over 20 minutes with torch backend. We will look into this issue and update you.

**Comment by mattdangerw** [comment]
Have we tried testing this on Keras 3.10? We recently added cudnn support for lstm on torch https://github.com/keras-team/keras/pull/21135

**Comment by sonali-kumari1** [comment]
Hi @mattdangerw -

Yes, I have tested this on keras `3.10` and on a GPU setup with torch backend, it takes approximately 90 seconds per epoch, resulting in a total training time of over 30 minutes for 20 epochs. This is slightly higher than the training time observed with version `3.9.2`. Attaching [gist](https://colab.sandbox.google.com/gist/sonali-kumari1/4b24b83bc33e96dd8669ab141263970a/20902.ipynb) for your reference. Thanks!

**Comment by fhchl** [comment]
It seems like the LSTM cuDNN implementation always silently fails and defaults back to the keras LSTM implementation.

For example, running  the following while [removing the try-catch here](https://github.com/keras-team/keras/compare/master...fhchl:keras:master)

```python
import numpy as np
import keras
from keras import layers

keras.utils.set_random_seed(42)

X = np.random.normal(size=(1, 2, 1))
y = np.random.normal(size=(1, 1))

model = keras.Sequential([
    layers.LSTM(1, input_shape=(1, 1), return_sequences=True, use_cudnn=True),
    layers.LSTM(1, use_cudnn=True),
])

model.compile(optimizer=keras.optimizers.SGD(learning_rate=1.), loss='mse')

model.fit(X, y, epochs=1, verbose=1)
```

gives

```
can't convert cuda:0 device type tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.
```

Could it be that there are also other problems with that implementation? For example, don't interfere [these detach calls here](https://github.com/keras-team/keras/blob/19ca9c1d7e02410acb353de8748f7cb9240dad30/keras/src/backend/torch/rnn.py#L700) with backprop?

PS: I tried to set a colab up, but couldn't figure out how to install keras from that fork into the colab environment ...

**Comment by spartanjoax** [comment]
I am experiencing the same. Is there any update on this issue?

**Comment by MarcosAsh** [comment]
Hi, this should be fixed by #22874: fhchl's  diagnosis was right torch._VF.lstm's return tuple was being unpacked wrong, raising an exception that the try/except around lstm()  swallowed, so every LSTM(use_cudnn='auto') silently fell back to the pure-Python loop. The fix restores the cuDNN path (verified on A100: a 2-layer BiLSTM drops 391ms to 119.6ms/step with 8x fewer kernel launches).

**Comment by jeffcarp** [comment]
Great, thanks for the investigation and fix!
